tools/autograd/load_derivatives.py

```diff
@@ -85,11 +87,147 @@ def create_derivative(f: NativeFunction, formula: str, var_names: Tuple[str, ...
 
     return Derivative(
         formula=formula,
+        original_formula=original_formula,
         var_names=var_names,
         saved_inputs=saved_inputs,
         saved_outputs=saved_outputs,
     )
 
+def create_forward_derivative(f: NativeFunction, formula: str, names: Tuple[str, ...]) -> ForwardDerivative:
+    assert len(names) == 1, "Forward derivatives can define gradients for only one output at a time"
+    var_name = names[0]
+    var_type: Optional[Type] = None
+    for r in f.func.returns:
+        if r.name == var_name:
+            var_type = r.type
+            break
+    # Handle default return names
+    if var_type is None:
+        if var_name == "result":
+            assert len(f.func.returns) == 1
+            var_type = f.func.returns[0].type
+        else:
+            res = re.findall(r"^result(\d+)$", var_name)
+            if len(res) == 1:
+                arg_idx = int(res[0])
+                var_type = f.func.returns[arg_idx].type
+
+    assert var_type is not None, "No matching output for forward derivative definition"
+    return ForwardDerivative(
+        formula=formula,
+        var_name=var_name,
+        var_type=var_type,
+        required_inputs_fw_grad=None,
+        required_inputs_primal=None)
+
+def postprocess_forward_derivatives(
+    f: NativeFunction,
+    defn_name: str,
+    all_arg_names: List[str],
+    derivatives: List[Derivative],
+    forward_derivatives: List[ForwardDerivative],
+    args_with_derivatives: Sequence[Binding]
+) -> List[ForwardDerivative]:
+
+    def find_required_inputs(formula: str, postfix: str) -> Tuple[str, ...]:
+        required_inputs = set()
+        for arg in args_with_derivatives:
+            if arg.type == 'TensorList':
+                # The functions taking TensorList handle everything internally
+                continue
+            arg_name = arg.name
+
+            found = re.search(IDENT_REGEX.format(arg_name), formula)
+            if found:
+                raise RuntimeError(f"The forward formula for {defn_name} is using the base name of the {arg_name} "
+                                   f"argument which is ambiguous. You should use {arg_name}_p to access the primal "
+                                   f"value and {arg_name}_t to access the tangent.")
+
+            found = re.search(IDENT_REGEX.format(arg_name + postfix), formula)
+            if found:
+                required_inputs.add(arg_name)
+
+        return tuple(required_inputs)
+
+    updated_derivatives: List[ForwardDerivative] = []
+
+    for defn in forward_derivatives:
+        formula = defn.formula
+        required_inputs_tangent = find_required_inputs(formula, "_t")
+        if formula == "auto_element_wise":
```

Comment:
Would be a ton easier to understand the transformation here with just a simple "before" and "after"